File: services/academic-service/src/test/java/com/visor/school/academic/model/ClassTest.java
package com.visor.school.academic.model;

import org.junit.jupiter.api.Test;
import java.time.LocalDate;
import java.util.UUID;
import static org.junit.jupiter.api.Assertions.*;

class ClassTest {
    @Test
    void shouldCreateHomeroomClassForGrades1To6() {
        Class homeroomClass = new Class("Grade 3 Homeroom", ClassType.HOMEROOM, null, 3, UUID.randomUUID(), null, "2024-2025", Term.FIRST_TERM, null, null, LocalDate.of(2024, 9, 1), null, ClassStatus.SCHEDULED);
        assertEquals(ClassType.HOMEROOM, homeroomClass.getClassType());
        assertEquals(3, homeroomClass.getGradeLevel());
        assertNotNull(homeroomClass.getHomeroomTeacherId());
        assertNull(homeroomClass.getSubject());
        assertNull(homeroomClass.getClassTeacherId());
    }
    @Test
    void shouldThrowExceptionForHomeroomClassWithGradeAbove6() {
        assertThrows(IllegalArgumentException.class, () -> new Class("Invalid Homeroom", ClassType.HOMEROOM, null, 7, UUID.randomUUID(), null, "2024-2025", Term.FIRST_TERM, null, null, LocalDate.of(2024, 9, 1), null, ClassStatus.SCHEDULED));
    }
    @Test
    void shouldThrowExceptionForHomeroomClassWithSubject() {
        assertThrows(IllegalArgumentException.class, () -> new Class("Invalid Homeroom", ClassType.HOMEROOM, "Mathematics", 3, UUID.randomUUID(), null, "2024-2025", Term.FIRST_TERM, null, null, LocalDate.of(2024, 9, 1), null, ClassStatus.SCHEDULED));
    }
    @Test
    void shouldThrowExceptionForHomeroomClassWithClassTeacherId() {
        assertThrows(IllegalArgumentException.class, () -> new Class("Invalid Homeroom", ClassType.HOMEROOM, null, 3, UUID.randomUUID(), UUID.randomUUID(), "2024-2025", Term.FIRST_TERM, null, null, LocalDate.of(2024, 9, 1), null, ClassStatus.SCHEDULED));
    }
    @Test
    void shouldCreateSubjectClassForAllGrades() {
        Class subjectClass = new Class("Mathematics 101", ClassType.SUBJECT, "Mathematics", 10, null, null, "2024-2025", Term.FIRST_TERM, null, null, LocalDate.of(2024, 9, 1), null, ClassStatus.SCHEDULED);
        assertEquals(ClassType.SUBJECT, subjectClass.getClassType());
        assertEquals("Mathematics", subjectClass.getSubject());
        assertEquals(10, subjectClass.getGradeLevel());
        assertNull(subjectClass.getHomeroomTeacherId());
    }
    @Test
    void shouldCreateSubjectClassForGrades1To6() {
        Class subjectClass = new Class("Art Grade 3", ClassType.SUBJECT, "Art", 3, null, null, "2024-2025", Term.FIRST_TERM, null, null, LocalDate.of(2024, 9, 1), null, ClassStatus.SCHEDULED);
        assertEquals(ClassType.SUBJECT, subjectClass.getClassType());
        assertEquals(3, subjectClass.getGradeLevel());
    }
    @Test
    void shouldThrowExceptionForSubjectClassWithoutSubject() {
        assertThrows(IllegalArgumentException.class, () -> new Class("Invalid Subject", ClassType.SUBJECT, null, 10, null, null, "2024-2025", Term.FIRST_TERM, null, null, LocalDate.of(2024, 9, 1), null, ClassStatus.SCHEDULED));
    }
    @Test
    void shouldThrowExceptionForSubjectClassWithHomeroomTeacherId() {
        assertThrows(IllegalArgumentException.class, () -> new Class("Invalid Subject", ClassType.SUBJECT, "Mathematics", 10, UUID.randomUUID(), null, "2024-2025", Term.FIRST_TERM, null, null, LocalDate.of(2024, 9, 1), null, ClassStatus.SCHEDULED));
    }
    @Test
    void shouldThrowExceptionForGradeLevelBelow1() {
        assertThrows(IllegalArgumentException.class, () -> new Class("Invalid Grade", ClassType.SUBJECT, "Mathematics", 0, null, null, "2024-2025", Term.FIRST_TERM, null, null, LocalDate.of(2024, 9, 1), null, ClassStatus.SCHEDULED));
    }
    @Test
    void shouldThrowExceptionForGradeLevelAbove12() {
        assertThrows(IllegalArgumentException.class, () -> new Class("Invalid Grade", ClassType.SUBJECT, "Mathematics", 13, null, null, "2024-2025", Term.FIRST_TERM, null, null, LocalDate.of(2024, 9, 1), null, ClassStatus.SCHEDULED));
    }
    @Test
    void shouldUpdateClassStatus() throws InterruptedException {
        Class classEntity = new Class("Test Class", ClassType.SUBJECT, "Mathematics", 5, null, null, "2024-2025", Term.FIRST_TERM, null, null, LocalDate.of(2024, 9, 1), null, ClassStatus.SCHEDULED);
        assertEquals(ClassStatus.SCHEDULED, classEntity.getStatus());
        var initialUpdatedAt = classEntity.getUpdatedAt();
        Thread.sleep(10);
        classEntity.updateStatus(ClassStatus.IN_PROGRESS);
        assertEquals(ClassStatus.IN_PROGRESS, classEntity.getStatus());
        assertTrue(classEntity.getUpdatedAt().isAfter(initialUpdatedAt));
    }
    @Test
    void shouldIncrementAndDecrementEnrollment() {
        Class classEntity = new Class("Test Class", ClassType.SUBJECT, "Mathematics", 5, null, null, "2024-2025", Term.FIRST_TERM, null, null, LocalDate.of(2024, 9, 1), null, ClassStatus.SCHEDULED);
        assertEquals(0, classEntity.getCurrentEnrollment());
        classEntity.incrementEnrollment();
        assertEquals(1, classEntity.getCurrentEnrollment());
        classEntity.incrementEnrollment();
        assertEquals(2, classEntity.getCurrentEnrollment());
        classEntity.decrementEnrollment();
        assertEquals(1, classEntity.getCurrentEnrollment());
    }
    @Test
    void shouldThrowExceptionWhenDecrementingEnrollmentBelowZero() {
        Class classEntity = new Class("Test Class", ClassType.SUBJECT, "Mathematics", 5, null, null, "2024-2025", Term.FIRST_TERM, null, null, LocalDate.of(2024, 9, 1), null, ClassStatus.SCHEDULED);
        assertThrows(IllegalArgumentException.class, classEntity::decrementEnrollment);
    }
}

File: services/academic-service/src/test/kotlin/com/visor/school/academicservice/model/ClassTest.kt
package com.visor.school.academicservice.model

import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import java.time.LocalDate
import java.util.UUID

class ClassTest {

    @Test
    fun `should create homeroom class for grades 1-6`() {
        val homeroomClass = Class(
            className = "Grade 3 Homeroom",
            classType = ClassType.HOMEROOM,
            subject = null,
            gradeLevel = 3,
            homeroomTeacherId = UUID.randomUUID(),
            classTeacherId = null,
            academicYear = "2024-2025",
            term = Term.FIRST_TERM,
            startDate = LocalDate.of(2024, 9, 1)
        )

        assertEquals(ClassType.HOMEROOM, homeroomClass.classType)
        assertEquals(3, homeroomClass.gradeLevel)
        assertNotNull(homeroomClass.homeroomTeacherId)
        assertNull(homeroomClass.subject)
        assertNull(homeroomClass.classTeacherId)
    }

    @Test
    fun `should throw exception for homeroom class with grade above 6`() {
        assertThrows<IllegalArgumentException> {
            Class(
                className = "Invalid Homeroom",
                classType = ClassType.HOMEROOM,
                subject = null,
                gradeLevel = 7,
                homeroomTeacherId = UUID.randomUUID(),
                classTeacherId = null,
                academicYear = "2024-2025",
                term = Term.FIRST_TERM,
                startDate = LocalDate.of(2024, 9, 1)
            )
        }
    }

    @Test
    fun `should throw exception for homeroom class with subject`() {
        assertThrows<IllegalArgumentException> {
            Class(
                className = "Invalid Homeroom",
                classType = ClassType.HOMEROOM,
                subject = "Mathematics", // Should be null for homeroom
                gradeLevel = 3,
                homeroomTeacherId = UUID.randomUUID(),
                classTeacherId = null,
                academicYear = "2024-2025",
                term = Term.FIRST_TERM,
                startDate = LocalDate.of(2024, 9, 1)
            )
        }
    }

    @Test
    fun `should throw exception for homeroom class with classTeacherId`() {
        assertThrows<IllegalArgumentException> {
            Class(
                className = "Invalid Homeroom",
                classType = ClassType.HOMEROOM,
                subject = null,
                gradeLevel = 3,
                homeroomTeacherId = UUID.randomUUID(),
                classTeacherId = UUID.randomUUID(), // Should be null for homeroom
                academicYear = "2024-2025",
                term = Term.FIRST_TERM,
                startDate = LocalDate.of(2024, 9, 1)
            )
        }
    }

    @Test
    fun `should create subject class for all grades`() {
        val subjectClass = Class(
            className = "Mathematics 101",
            classType = ClassType.SUBJECT,
            subject = "Mathematics",
            gradeLevel = 10,
            homeroomTeacherId = null,
            classTeacherId = null,
            academicYear = "2024-2025",
            term = Term.FIRST_TERM,
            startDate = LocalDate.of(2024, 9, 1)
        )

        assertEquals(ClassType.SUBJECT, subjectClass.classType)
        assertEquals("Mathematics", subjectClass.subject)
        assertEquals(10, subjectClass.gradeLevel)
        assertNull(subjectClass.homeroomTeacherId)
    }

    @Test
    fun `should create subject class for grades 1-6`() {
        val subjectClass = Class(
            className = "Art Grade 3",
            classType = ClassType.SUBJECT,
            subject = "Art",
            gradeLevel = 3,
            homeroomTeacherId = null,
            classTeacherId = null,
            academicYear = "2024-2025",
            term = Term.FIRST_TERM,
            startDate = LocalDate.of(2024, 9, 1)
        )

        assertEquals(ClassType.SUBJECT, subjectClass.classType)
        assertEquals(3, subjectClass.gradeLevel)
    }

    @Test
    fun `should throw exception for subject class without subject`() {
        assertThrows<IllegalArgumentException> {
            Class(
                className = "Invalid Subject",
                classType = ClassType.SUBJECT,
                subject = null, // Required for subject classes
                gradeLevel = 10,
                homeroomTeacherId = null,
                classTeacherId = null,
                academicYear = "2024-2025",
                term = Term.FIRST_TERM,
                startDate = LocalDate.of(2024, 9, 1)
            )
        }
    }

    @Test
    fun `should throw exception for subject class with homeroomTeacherId`() {
        assertThrows<IllegalArgumentException> {
            Class(
                className = "Invalid Subject",
                classType = ClassType.SUBJECT,
                subject = "Mathematics",
                gradeLevel = 10,
                homeroomTeacherId = UUID.randomUUID(), // Should be null for subject classes
                classTeacherId = null,
                academicYear = "2024-2025",
                term = Term.FIRST_TERM,
                startDate = LocalDate.of(2024, 9, 1)
            )
        }
    }

    @Test
    fun `should throw exception for grade level below 1`() {
        assertThrows<IllegalArgumentException> {
            Class(
                className = "Invalid Grade",
                classType = ClassType.SUBJECT,
                subject = "Mathematics",
                gradeLevel = 0, // Invalid
                homeroomTeacherId = null,
                classTeacherId = null,
                academicYear = "2024-2025",
                term = Term.FIRST_TERM,
                startDate = LocalDate.of(2024, 9, 1)
            )
        }
    }

    @Test
    fun `should throw exception for grade level above 12`() {
        assertThrows<IllegalArgumentException> {
            Class(
                className = "Invalid Grade",
                classType = ClassType.SUBJECT,
                subject = "Mathematics",
                gradeLevel = 13, // Invalid
                homeroomTeacherId = null,
                classTeacherId = null,
                academicYear = "2024-2025",
                term = Term.FIRST_TERM,
                startDate = LocalDate.of(2024, 9, 1)
            )
        }
    }

    @Test
    fun `should update class status`() {
        val classEntity = Class(
            className = "Test Class",
            classType = ClassType.SUBJECT,
            subject = "Mathematics",
            gradeLevel = 5,
            homeroomTeacherId = null,
            classTeacherId = null,
            academicYear = "2024-2025",
            term = Term.FIRST_TERM,
            startDate = LocalDate.of(2024, 9, 1),
            status = ClassStatus.SCHEDULED
        )

        assertEquals(ClassStatus.SCHEDULED, classEntity.status)
        val initialUpdatedAt = classEntity.updatedAt

        Thread.sleep(10)
        classEntity.updateStatus(ClassStatus.IN_PROGRESS)

        assertEquals(ClassStatus.IN_PROGRESS, classEntity.status)
        assertTrue(classEntity.updatedAt.isAfter(initialUpdatedAt))
    }

    @Test
    fun `should increment and decrement enrollment`() {
        val classEntity = Class(
            className = "Test Class",
            classType = ClassType.SUBJECT,
            subject = "Mathematics",
            gradeLevel = 5,
            homeroomTeacherId = null,
            classTeacherId = null,
            academicYear = "2024-2025",
            term = Term.FIRST_TERM,
            startDate = LocalDate.of(2024, 9, 1),
            maxCapacity = 30
        )

        assertEquals(0, classEntity.currentEnrollment)

        classEntity.incrementEnrollment()
        assertEquals(1, classEntity.currentEnrollment)

        classEntity.incrementEnrollment()
        assertEquals(2, classEntity.currentEnrollment)

        classEntity.decrementEnrollment()
        assertEquals(1, classEntity.currentEnrollment)
    }

    @Test
    fun `should throw exception when decrementing enrollment below zero`() {
        val classEntity = Class(
            className = "Test Class",
            classType = ClassType.SUBJECT,
            subject = "Mathematics",
            gradeLevel = 5,
            homeroomTeacherId = null,
            classTeacherId = null,
            academicYear = "2024-2025",
            term = Term.FIRST_TERM,
            startDate = LocalDate.of(2024, 9, 1)
        )

        assertThrows<IllegalArgumentException> {
            classEntity.decrementEnrollment()
        }
    }
}

